### PR TITLE
Pytest benchmarks

### PR DESCRIPTION
**Pull request by SaitejaUtpala**
benchmarks infra

### CODE REVIEW

**Comment by SaitejaUtpala** [comment]
@ninamiolane  I am skipping some of tests in 32 bit backends in tensorflow and pytorch. They keep failing every now and then. Thanks!

File: benchmarks/exp/time_exp.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import pytest
+
+
+def read_benchmark_exp_data():
+    """Read benchmark parameters from file benchmark_params.pkl."""
+    data = []
+    ids = []
+    df = pd.read_pickle("./benchmark_params.pkl")
+    params = list(df.itertuples(index=False))
+    print(params)
```
Comment (ninamiolane) [inline benchmarks/exp/time_exp.py:15]: Is it a leftover print? maybe use logging if not?

File: benchmarks/exp/time_exp.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+        module = import_module(module)
+        manifold = getattr(module, manifold)(*manifold_args)
+        metric = getattr(module, metric)(*metric_args)
+        base_point = manifold.random_point(n_samples)
+        tangent_vec = manifold.random_tangent_vec(n_samples, base_point)
+        exp_args = (tangent_vec, base_point)
+        data.append((metric, exp_args, exp_kwargs))
+
+    return (data, ids)
```
Comment (ninamiolane) [inline benchmarks/exp/time_exp.py:37]: not sure you need (, ) ?

File: benchmarks/script.sh [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+done
+echo "Manifold: $manifold";
+echo "N Samples: $n_samples";
+echo "Operation: $operation";
+
+
+python generate_benchmark_params.py
+pytest exp/time_exp.py  --benchmark-columns='min, max'  --benchmark-sort='fullname'
+pytest log/time_log.py  --benchmark-columns='min, max'  --benchmark-sort='fullname'
+pytest dist/time_dist.py --benchmark-columns='min, max'  --benchmark-sort='fullname'
```
Comment (ninamiolane) [inline benchmarks/script.sh:22]: I don't see the dist folder, will this fail?

File: benchmarks/exp/time_exp.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+
+benchmark_data, benchmark_ids = read_benchmark_exp_data()
+
+
+@pytest.mark.parametrize(
+    "metric, exp_args, exp_kwargs", benchmark_data, ids=benchmark_ids
+)
+def test_benchmark_exp(metric, exp_args, exp_kwargs, benchmark):
```
Comment (ninamiolane) [inline benchmarks/exp/time_exp.py:46]: Is this the test of the benchmark or the actual benchmark? I'm thinking it's the second, and pytest wants `test_` at the beginning of the function's name?

Could you also add the parameters' description there?

File: benchmarks/generate_benchmark_params.py [hunk 5]
```diff
@@ -0,0 +1,521 @@
+"""Benchmarking parameters generation file."""
+
+import argparse
+from itertools import chain, product
+
+import pandas as pd
+
+parser = argparse.ArgumentParser(
```
Comment (ninamiolane) [inline benchmarks/generate_benchmark_params.py:8]: What are the pros/cons of using argument parser as opposed to a config file?

I personally find config.yml file more intuitive and readable, but I wonder what the community recommends there.

File: benchmarks/generate_benchmark_params.py [hunk 6]
```diff
@@ -0,0 +1,521 @@
...
+)
+args = parser.parse_args()
+
+
+def spd_manifold_params(n_samples):
+    """Generate spd manifold benchmarking parameters.
+
+    Parameters
+    ----------
+        n_samples : int
```
Comment (ninamiolane) [inline benchmarks/generate_benchmark_params.py:34]: bad indent

File: benchmarks/generate_benchmark_params.py [hunk 7]
```diff
@@ -0,0 +1,521 @@
...
+            Number of samples to be used.
+
+    Returns
+    -------
+        _:list.
```
Comment (ninamiolane) [inline benchmarks/generate_benchmark_params.py:39]: bad indent

File: benchmarks/generate_benchmark_params.py [hunk 8]
```diff
@@ -0,0 +1,521 @@
...
+    def affine_metric_params():
+        params = []
+        metric = "SPDMetricAffine"
+        power_args = [-0.5, 1, 0.5]
+        metric_args = list(product([item for item, in manifold_args], power_args))
+        manifold_args_re = [
+            item for item in manifold_args for i in range(len(power_args))
+        ]
+        kwargs = {}
+        common = (manifold, module, metric, n_samples, kwargs)
```
Comment (ninamiolane) [inline benchmarks/generate_benchmark_params.py:55]: I dont think that you need parentheses here

File: benchmarks/generate_benchmark_params.py [hunk 9]
```diff
@@ -0,0 +1,521 @@
...
+
+    def log_euclidean_metric_params():
+        params = []
+        metric = "SPDMetricLogEuclidean"
+        metric_args = manifold_args
+        kwargs = {}
+        common = (manifold, module, metric, n_samples, kwargs)
+        for manifold_arg, metric_arg in zip(manifold_args, metric_args):
+            params += [common + (manifold_arg, metric_arg)]
+        return params
```
Comment (ninamiolane) [inline benchmarks/generate_benchmark_params.py:92]: The lines:
```
        kwargs = {}
        common = (manifold, module, metric, n_samples, kwargs)
        for manifold_arg, metric_arg in zip(manifold_args, metric_args):
            params += [common + (manifold_arg, metric_arg)]
        return params
```
seems to be repeated for the params of each metric. 

Should they be factorized? (maybe not, maybe it is more readable/editable this way?)

File: benchmarks/generate_benchmark_params.py [hunk 10]
```diff
@@ -0,0 +1,521 @@
...
+        )
+    )
+
+
+def stiefel_params(n_samples):
+    """Generate stiefel benchmarking parameters.
+
+    Parameters
+    ----------
+        n_samples : int
```
Comment (ninamiolane) [inline benchmarks/generate_benchmark_params.py:111]: bad indent (check all other indents in this file)

File: benchmarks/generate_benchmark_params.py [hunk 11]
```diff
@@ -0,0 +1,521 @@
...
+    Returns
+    -------
+        _:list.
+        List of params.
+    """
+    manifold = "SPDMatrices"
+    manifold_args = [(2,), (5,), (10,)]
+    module = "geomstats.geometry.spd_matrices"
+
+    def affine_metric_params():
```
Comment (ninamiolane) [inline benchmarks/generate_benchmark_params.py:46]: Need to be `spd_affine_metric_params` for consistency with the rest of this file:

e.g. StiefelCanonicalMetric became stiefel_canonical_metric

The rule to use exactly the name of the metric seems good to me.

File: benchmarks/generate_benchmark_params.py [hunk 12]
```diff
@@ -0,0 +1,521 @@
...
+    Returns
+    -------
+        _:list.
+        List of params.
+    """
+    manifold = "FullRankCorrelationMatrices"
+    manifold_args = [(3,), (5,)]
+    module = "geomstats.geometry.full_rank_correlation_matrices"
+
+    def full_rank_correlation_matrices_metric_params():
```
Comment (ninamiolane) [inline benchmarks/generate_benchmark_params.py:333]: change name, it show follow exactly the name of corresponding metric

File: benchmarks/log/time_log.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+import pytest
+
+
+def read_benchmark_log_data():
+    """Read benchmark parameters from file benchmark_params.pkl."""
+    data = []
+    ids = []
+    df = pd.read_pickle("benchmark_params.pkl")
+    params = list(df.itertuples(index=False))
+    print(params)
```
Comment (ninamiolane) [inline benchmarks/log/time_log.py:15]: Is it a leftover print? maybe use logging if not?

**Review by ninamiolane — APPROVED** [review]
Very nice!! Congrats on getting this to work. Good to merge as soon as my comments are addressed.

File: benchmarks/generate_benchmark_params.py (hunk 9 above)
Comment (SaitejaUtpala) [reply benchmarks/generate_benchmark_params.py:92]: yup.. they could be factorized. I didn't want to over simplify things without knowing completely. Let's say if some metric requires different  processing from information geometry.

File: benchmarks/generate_benchmark_params.py (hunk 5 above)
Comment (SaitejaUtpala) [reply benchmarks/generate_benchmark_params.py:8]: Either could be fine. for example pytorch uses command line args. Happy to change it if you prefer otherwise. Thanks!

File: benchmarks/script.sh (hunk 3 above)
Comment (SaitejaUtpala) [reply benchmarks/script.sh:22]: @ninamiolane  sorry. turns out .gitignore had this dist/ folder. So folder got ignored. completely. So let me what it should be done here. Thanks!

**Comment by SaitejaUtpala** [comment]
@ninamiolane  Thanks for the review. Addressed your review comments

File: benchmarks/generate_benchmark_params.py [hunk 14]
```diff
@@ -0,0 +1,521 @@
...
+        manifold_args_re = [
+            item for item in manifold_args for i in range(len(power_args))
+        ]
+        kwargs = {}
+        common = manifold, module, metric, n_samples, kwargs
+        for manifold_arg, metric_arg in zip(manifold_args_re, metric_args):
+            params += [common + (manifold_arg, metric_arg)]
+        return params
+
+    def bures_wasserstein_metric_params():
```
Comment (ninamiolane) [inline benchmarks/generate_benchmark_params.py:60]: spd_bures_wasserstein, look at other names in this function too

File: benchmarks/dist/time_dist.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    """Benchmark metric distance map.
+
+    Parameters
+    ----------
+    metric : object
+        Metric object.
+    dist_args : tuple
+        Arguments to dist function.
+    dist_kwargs : tuple
+        Keyword argumetns to dist function.
```
Comment (ninamiolane) [inline benchmarks/dist/time_dist.py:55]: typo

File: benchmarks/exp/time_exp.py [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    """Benchmark metric exponential map.
+
+    Parameters
+    ----------
+    metric : object
+        Metric object.
+    exp_args : tuple
+        Arguments to exp function.
+    exp_kwargs : tuple
+        Keyword argumetns to exp function.
```
Comment (ninamiolane) [inline benchmarks/exp/time_exp.py:55]: typo

File: benchmarks/log/time_log.py [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+    """Benchmark metric logarithm map.
+
+    Parameters
+    ----------
+    metric : object
+        Metric object.
+    log_args : tuple
+        Arguments to log function.
+    log_kwargs : tuple
+        Keyword argumetns to log function.
```
Comment (ninamiolane) [inline benchmarks/log/time_log.py:55]: typo

**Review by ninamiolane — APPROVED** [review]
Look fo rTypo + func names in SPD
